### PR TITLE
Add tests to publisher and init modules of rcl

### PR DESCRIPTION
**Pull request by Blast545**
As the title says, pushing to increase the coverage of the `rcl` module.

The tests added to loan functions won't increase coverage at the moment. I'd like a bit of extra attention there, not sure if that's the correct use of the API.

### CODE REVIEW

File: rcl/test/rcl/test_init.cpp [hunk 1]
```diff
@@ -271,3 +271,17 @@ TEST_F(CLASSNAME(TestRCLFixture, RMW_IMPLEMENTATION), test_rcl_get_instance_id)
   ret = rcl_context_fini(&context);
   EXPECT_EQ(ret, RCL_RET_OK);
 }
+
+TEST_F(CLASSNAME(TestRCLFixture, RMW_IMPLEMENTATION), test_rcl_init_options_access) {
+  rcl_ret_t ret;
+  rcl_init_options_t init_options = rcl_get_zero_initialized_init_options();
+  ret = rcl_init_options_init(&init_options, rcl_get_default_allocator());
```
Comment (hidmic) [inline rcl/test/rcl/test_init.cpp:278]: @Blast545 nit: consider collapsing `ret` declaration (line 276) and definition (line 278, this line) into a single statement.

File: rcl/test/rcl/test_init.cpp [hunk 2]
```diff
@@ -271,3 +271,17 @@ TEST_F(CLASSNAME(TestRCLFixture, RMW_IMPLEMENTATION), test_rcl_get_instance_id)
...
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+  rmw_init_options_t * options = rcl_init_options_get_rmw_init_options(&init_options);
+  EXPECT_EQ(0u, options->instance_id);
```
Comment (hidmic) [inline rcl/test/rcl/test_init.cpp:282]: @Blast545 consider asserting that `options != nullptr` before using it.

File: rcl/test/rcl/test_publisher.cpp [hunk 3]
```diff
@@ -290,3 +290,93 @@ TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_init_
...
 }
+
+TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_loan) {
+  rcl_ret_t ret;
+  rcl_publisher_t publisher = rcl_get_zero_initialized_publisher();
+  const rosidl_message_type_support_t * ts =
+    ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, Strings);
+  const char * topic_name = "chatter";
+  rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+  ret = rcl_publisher_init(&publisher, this->node_ptr, ts, topic_name, &publisher_options);
```
Comment (hidmic) [inline rcl/test/rcl/test_publisher.cpp:301]: @Blast545 same nit as above about `ret`.

File: rcl/test/rcl/test_publisher.cpp [hunk 4]
```diff
@@ -290,3 +290,93 @@ TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_init_
...
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  });
+
+  test_msgs__msg__Strings msg_loaned;
+  if (rcl_publisher_can_loan_messages(&publisher)) {
+    EXPECT_EQ(
+      RCL_RET_OK, rcl_borrow_loaned_message(
+        &publisher,
+        ts,
+        reinterpret_cast<void **>(&msg_loaned)));
```
Comment (hidmic) [inline rcl/test/rcl/test_publisher.cpp:315]: @Blast545 hmm, I have the impression that `rcl_borrow_loaned_message()` returns a pointer to an allocated (loaned) message, in which case `msg_loaned` should be a pointer to `test_msgs__msg__Strings` and this cast shouldn't be necessary (implicit casts to `void` pointers are allowed)

File: rcl/test/rcl/test_publisher.cpp [hunk 5]
```diff
@@ -290,3 +290,93 @@ TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_init_
...
+    ASSERT_TRUE(rosidl_runtime_c__String__assign(&msg_loaned.string_value, "testing"));
+    EXPECT_EQ(
+      RCL_RET_OK, rcl_publish_loaned_message(
+        &publisher,
+        reinterpret_cast<void *>(&msg_loaned),
```
Comment (hidmic) [inline rcl/test/rcl/test_publisher.cpp:320]: @Blast545 same about the cast.

File: rcl/test/rcl/test_publisher.cpp [hunk 6]
```diff
@@ -290,3 +290,93 @@ TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_init_
...
+}
+
+TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_access_functions) {
+  rcl_ret_t ret;
+  rcl_publisher_t publisher = rcl_get_zero_initialized_publisher();
+  const rosidl_message_type_support_t * ts =
+    ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, Strings);
+  const char * topic_name = "chatter";
+  rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+  ret = rcl_publisher_init(&publisher, this->node_ptr, ts, topic_name, &publisher_options);
```
Comment (hidmic) [inline rcl/test/rcl/test_publisher.cpp:332]: @Blast545 same nit about `rcl_ret_t`

File: rcl/test/rcl/test_publisher.cpp [hunk 7]
```diff
@@ -290,3 +290,93 @@ TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_init_
...
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    rcl_ret_t ret = rcl_publisher_fini(&publisher, this->node_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  });
+
+  const rcl_publisher_options_t * publisher_options_rcv;
+  publisher_options_rcv = rcl_publisher_get_options(&publisher);
```
Comment (hidmic) [inline rcl/test/rcl/test_publisher.cpp:341]: @Blast545 nit: the last two lines can be collapsed into one.

Comment (hidmic) [reply rcl/test/rcl/test_publisher.cpp:341]: And an assert for `publisher_options_rcv != nullptr` should occur before using it.

File: rcl/test/rcl/test_publisher.cpp [hunk 8]
```diff
@@ -290,3 +290,93 @@ TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_init_
...
+}
+
+TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_assert_lieveliness) {
+  rcl_ret_t ret;
+  rcl_publisher_t publisher = rcl_get_zero_initialized_publisher();
+  const rosidl_message_type_support_t * ts =
+    ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, Strings);
+  const char * topic_name = "chatter";
+  rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+  ret = rcl_publisher_init(&publisher, this->node_ptr, ts, topic_name, &publisher_options);
```
Comment (hidmic) [inline rcl/test/rcl/test_publisher.cpp:372]: @Blast545 nit: same comment about `rcl_ret_t`.

File: rcl/test/rcl/test_publisher.cpp (hunk 4 above)
Comment (Blast545) [reply rcl/test/rcl/test_publisher.cpp]: I fixed `msg_loaned` to be a pointer to `test_msgs__msg__Strings`, and actually added a pointer to that one for this part. However, I still got:
`invalid conversion from ‘test_msgs__msg__Strings***’ to ‘void**’ [-fpermissive]`
and had to left the cast it in my last commit.

Comment (hidmic) [reply rcl/test/rcl/test_publisher.cpp]: Hmm, why do you provide a pointer to your pointer to pointer? `msg_loaned_ptr` should be enough, no leading `&`

File: rcl/test/rcl/test_publisher.cpp [hunk 9]
```diff
@@ -292,53 +292,54 @@ TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_init_
...
+    rcl_publisher_init(&publisher, this->node_ptr, ts, topic_name, &publisher_options);
   ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
   OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
   {
     rcl_ret_t ret = rcl_publisher_fini(&publisher, this->node_ptr);
     EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
   });
 
-  test_msgs__msg__Strings msg_loaned;
+  test_msgs__msg__Strings * msg_loaned;
```
Comment (hidmic) [inline rcl/test/rcl/test_publisher.cpp:309]: @Blast545 you may want initialize this to `nullptr`.

File: rcl/test/rcl/test_publisher.cpp [hunk 10]
```diff
@@ -292,53 +292,54 @@ TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_init_
...
         ts,
-        reinterpret_cast<void **>(&msg_loaned)));
-    ASSERT_TRUE(rosidl_runtime_c__String__assign(&msg_loaned.string_value, "testing"));
+        reinterpret_cast<void **>(&msg_loaned_ptr)));
+    ASSERT_TRUE(rosidl_runtime_c__String__assign(&(msg_loaned->string_value), "testing"));
     EXPECT_EQ(
       RCL_RET_OK, rcl_publish_loaned_message(
         &publisher,
-        reinterpret_cast<void *>(&msg_loaned),
+        &msg_loaned,
```
Comment (hidmic) [inline rcl/test/rcl/test_publisher.cpp:321]: Hmm, doesn't `rcl_publish_loaned_message()` take a pointer to message (instead of a pointer to pointer)?

File: rcl/test/rcl/test_publisher.cpp [hunk 11]
```diff
@@ -290,3 +290,94 @@ TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_init_
...
+  EXPECT_TRUE(rcutils_allocator_is_valid(&(publisher_options_rcv->allocator)));
+
+  rmw_publisher_t * pub_rmw_handle = rcl_publisher_get_rmw_handle(&publisher);
+  EXPECT_NE(nullptr, pub_rmw_handle);
+
+  rcl_context_t * pub_context = rcl_publisher_get_context(&publisher);
+  EXPECT_TRUE(rcl_context_is_valid(pub_context));
+  EXPECT_EQ(rcl_context_get_instance_id(context_ptr), rcl_context_get_instance_id(pub_context));
+
+  EXPECT_EQ(nullptr, rcl_publisher_get_options(nullptr));
```
Comment (brawner) [inline rcl/test/rcl/test_publisher.cpp:356]: I think it would be good make sure that all cases checked by `rcl_publisher_is_valid` result in the expected output in these functions. Besides nullptr, that would be `publisher->impl` is null and `publisher->impl->rmw_handle` is null, and publisher->impl->context is null.

File: rcl/test/rcl/test_publisher.cpp [hunk 12]
```diff
@@ -290,3 +290,94 @@ TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_init_
...
+    rcl_publisher_init(&publisher, this->node_ptr, ts, topic_name, &publisher_options);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    rcl_ret_t ret = rcl_publisher_fini(&publisher, this->node_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  });
+
+  EXPECT_EQ(RCL_RET_OK, rcl_publisher_assert_liveliness(&publisher));
+  EXPECT_EQ(RCL_RET_PUBLISHER_INVALID, rcl_publisher_assert_liveliness(nullptr));
```
Comment (brawner) [inline rcl/test/rcl/test_publisher.cpp:382]: Same comment.

**Comment by brawner** [comment]
Also can you run CI with coverage?

File: rcl/test/rcl/test_publisher.cpp (hunk 11 above)
Comment (Blast545) [reply rcl/test/rcl/test_publisher.cpp:356]: Given that all of these functions use internally `rcl_publisher_is_valid` or `rcl_publisher_is_valid_except_context` I think the best idea would be adding tests in particular for those two functions.

However, I think the second case you mention can't be achieved with use of the public API/definitions. Should I add the private part of the implementation to make that test?

**Comment by Blast545** [comment]
CI coverage:

Coverage: [Build Status](https://ci.ros2.org/job/ci_linux_coverage/108/)

**Comment by Blast545** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10841/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6214/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8832/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10729/)

File: rcl/test/rcl/test_init.cpp [hunk 13]
```diff
@@ -271,3 +271,17 @@ TEST_F(CLASSNAME(TestRCLFixture, RMW_IMPLEMENTATION), test_rcl_get_instance_id)
   ret = rcl_context_fini(&context);
   EXPECT_EQ(ret, RCL_RET_OK);
 }
+
+TEST_F(CLASSNAME(TestRCLFixture, RMW_IMPLEMENTATION), test_rcl_init_options_access) {
```
Comment (brawner) [inline rcl/test/rcl/test_init.cpp:275]: While you're editing this file, you can add two lines of coverage by adding a test elsewhere that runs test_init with a non-zero argc and a nullptr as the argv.

I think ultimately, we'll have to mock the functions `rcl_init` calls to get 100% of coverage though.

https://ci.ros2.org/job/ci_linux_coverage/115/cobertura/src_ros2_rcl_rcl_src_rcl/init_c/

File: rcl/test/rcl/test_publisher.cpp [hunk 14]
```diff
@@ -290,3 +290,99 @@ TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_init_
...
+  {
+    rcl_ret_t ret = rcl_publisher_fini(&publisher, this->node_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  });
+
+  test_msgs__msg__Strings * msg_loaned = nullptr;
+  test_msgs__msg__Strings ** msg_loaned_ptr = &msg_loaned;
+  if (rcl_publisher_can_loan_messages(&publisher)) {
+    EXPECT_EQ(
+      RCL_RET_OK, rcl_borrow_loaned_message(
```
[the remark this answers is not in the document]
Comment (hidmic) [reply rcl/test/rcl/test_publisher.cpp:313]: I believe no Tier 1 RMW implementation supports message loaning.

[the remark this answers is not in the document]
Comment (brawner) [reply rcl/test/rcl/test_publisher.cpp:313]: Oh got it. Thanks @hidmic.

File: rcl/test/rcl/test_publisher.cpp [hunk 15]
```diff
@@ -290,3 +290,99 @@ TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_init_
   EXPECT_EQ(RCL_RET_BAD_ALLOC, ret) << rcl_get_error_string().str;
   rcl_reset_error();
 }
+
```
[the remark this answers is not in the document]
Comment (Blast545) [reply rcl/test/rcl/test_publisher.cpp:295]: Added them. Although the loaned function messages I think are special cases to be discussed how to proceed.

File: rcl/test/rcl/test_init.cpp (hunk 13 above)
Comment (brawner) [reply rcl/test/rcl/test_init.cpp:275]: Was this added?

**Comment by brawner** [comment]
Rerunning coverage: [Build Status](https://ci.ros2.org/job/ci_linux_coverage/117/)

File: rcl/test/rcl/test_init.cpp (hunk 13 above)
Comment (Blast545) [reply rcl/test/rcl/test_init.cpp:284]: I added them before your check, but github is not working properly. I closed and reopened the PR and now they appear.

**Comment by Blast545** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10873/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6236/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8857/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10763/)

File: rcl/test/rcl/test_publisher.cpp [hunk 16]
```diff
@@ -290,3 +300,193 @@ TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_init_
...
+  rcl_context_t * pub_context = rcl_publisher_get_context(&publisher);
+  EXPECT_TRUE(rcl_context_is_valid(pub_context));
+  EXPECT_EQ(rcl_context_get_instance_id(context_ptr), rcl_context_get_instance_id(pub_context));
+
+  EXPECT_EQ(RCL_RET_OK, rcl_publisher_assert_liveliness(&publisher));
+
+  size_t count_size;
+  test_msgs__msg__BasicTypes msg;
+  rcl_serialized_message_t serialized_msg = rmw_get_zero_initialized_serialized_message();
+  rcl_publisher_impl_t * saved_impl = (&publisher)->impl;
```
Comment (hidmic) [inline rcl/test/rcl/test_publisher.cpp:371]: @Blast545 nit: why not
```suggestion
  rcl_publisher_impl_t * saved_impl = publisher.impl;
```
? Same below.

File: rcl/test/rcl/test_publisher.cpp [hunk 17]
```diff
@@ -290,3 +300,193 @@ TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_init_
...
+  EXPECT_NE(nullptr, rcl_publisher_get_topic_name(&publisher));
+  EXPECT_NE(nullptr, rcl_publisher_get_rmw_handle(&publisher));
+  EXPECT_NE(nullptr, rcl_publisher_get_actual_qos(&publisher));
+  EXPECT_NE(nullptr, rcl_publisher_get_options(&publisher));
+  rcl_reset_error();
+  EXPECT_FALSE(rcl_publisher_is_valid(&publisher));
+  rcl_reset_error();
+  EXPECT_EQ(nullptr, rcl_publisher_get_context(&publisher));
+  rcl_reset_error();
+  EXPECT_EQ(false, rcl_publisher_can_loan_messages(&publisher));
```
Comment (hidmic) [inline rcl/test/rcl/test_publisher.cpp:387]: @Blast545 nit: 
```suggestion
  EXPECT_FALSE(rcl_publisher_can_loan_messages(&publisher));
```

File: rcl/test/rcl/test_publisher.cpp [hunk 18]
```diff
@@ -290,3 +300,193 @@ TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_init_
...
+  rcl_reset_error();
+  EXPECT_EQ(nullptr, rcl_publisher_get_rmw_handle(&publisher));
+  rcl_reset_error();
+  EXPECT_EQ(nullptr, rcl_publisher_get_actual_qos(&publisher));
+  rcl_reset_error();
+  EXPECT_EQ(nullptr, rcl_publisher_get_options(&publisher));
+  rcl_reset_error();
+  EXPECT_EQ(nullptr, rcl_publisher_get_context(&publisher));
+  rcl_reset_error();
+  EXPECT_EQ(false, rcl_publisher_can_loan_messages(&publisher));
```
Comment (hidmic) [inline rcl/test/rcl/test_publisher.cpp:418]: @Blast545 same about using `EXPECT_FALSE` instead.

File: rcl/test/rcl/test_init.cpp [hunk 19]
```diff
@@ -271,3 +280,21 @@ TEST_F(CLASSNAME(TestRCLFixture, RMW_IMPLEMENTATION), test_rcl_get_instance_id)
...
+  rcl_init_options_t init_options = rcl_get_zero_initialized_init_options();
+  rcl_ret_t ret = rcl_init_options_init(&init_options, rcl_get_default_allocator());
```
Comment (hidmic) [inline rcl/test/rcl/test_init.cpp:286]: @Blast545 aren't we missing an `rcl_init_options_fini()` call somewhere in this test?

File: rcl/test/rcl/test_init.cpp [hunk 20]
```diff
@@ -271,3 +280,21 @@ TEST_F(CLASSNAME(TestRCLFixture, RMW_IMPLEMENTATION), test_rcl_get_instance_id)
...
+  rmw_init_options_t * options = rcl_init_options_get_rmw_init_options(&init_options);
+  ASSERT_NE(nullptr, options);
+  EXPECT_EQ(0u, options->instance_id);
+  EXPECT_EQ(nullptr, options->impl);
+
+  const rcl_allocator_t * options_allocator = rcl_init_options_get_allocator(&init_options);
+  EXPECT_TRUE(rcutils_allocator_is_valid(options_allocator));
+
+  rcl_init_options_t init_options_dst = rcl_get_zero_initialized_init_options();
+  EXPECT_EQ(RCL_RET_OK, rcl_init_options_copy(&init_options, &init_options_dst));
```
Comment (hidmic) [inline rcl/test/rcl/test_init.cpp:298]: @Blast545 same about an `rcl_init_options_fini()` call missing, this time for `init_options_dst`.

File: rcl/test/rcl/test_init.cpp (hunk 19 above)
Comment (Blast545) [reply rcl/test/rcl/test_init.cpp:286]: You are right! Added in the last commit

**Comment by Blast545** [comment]
Force pushed to rebase with master, CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10905/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6257/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8879/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10797/)

File: rcl/test/rcl/test_init.cpp [hunk 21]
```diff
@@ -271,3 +280,24 @@ TEST_F(CLASSNAME(TestRCLFixture, RMW_IMPLEMENTATION), test_rcl_get_instance_id)
...
+  EXPECT_EQ(RCL_RET_ALREADY_INIT, rcl_init_options_copy(&init_options, &init_options_dst));
+
+  EXPECT_EQ(RCL_RET_OK, rcl_init_options_fini(&init_options));
```
Comment (hidmic) [inline rcl/test/rcl/test_init.cpp:301]: @Blast545 what if the assertion in [line 290](https://github.com/ros2/rcl/pull/657/files#diff-c67929cc01d6cd07de8ab134e5d2cfd2R290) fails?

Comment (Blast545) [reply rcl/test/rcl/test_init.cpp:301]: AFAIK, the test will not continue if that `ASSERT_NE` fails. Am I missing something from your question?

Comment (hidmic) [reply rcl/test/rcl/test_init.cpp:301]: Correct, in which case this `rcl_init_options_fini()` call won't be executed.

Comment (Blast545) [reply rcl/test/rcl/test_init.cpp]: Oh, thanks! Totally missed that. Fixed in the last commit.

**Comment by Blast545** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10918/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6268/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8890/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10819/)
